### ISSUE TITLE
Interop with Non-NServiceBus Endpoints

### ISSUE THREAD

**Issue by gsogol** [Particular/NServiceBus.RabbitMQ#13]
When a non-NSB endpoint ((e.g. Spring Integration/Java, etc) needs to publish a message to a NServiceBus endpoint, .Net engineers force non-NSB producers to add NSB-specific semantics:

- producers have to make sure to properly add a .Net namespace Company.Package.SubPackage.

- due to NSB's serialization, objects in the XML have to be in Camel Case instead of language's (java, python, whatever) own semantics.

- message type had to be added to the message header with the exact same
name as the Exchange. For example,  Company.Package.SubPackage.Events.QuoteRequested is the XML
object we send and the Exchange and had to be defined as the Message Type
in the message header (see attached image)

- content_type had to be set to text/xml. A language like Java is defaulting to
octet-stream. Potentially, telling NServiceBus to treat X as though it's Y can solve an issue.

- CorrelationID needs to be set to the same value as the MessageID in
the message header. No reply is being expected. Java producers are confused why they need such a header. Again, potentially, telling NSB to treat some property as the CorrelationID would suffice. Producers can decide what they want to name things and we just need to tell NSB to treat some header as CorrelationID.

In summary, NServiceBus needs to be flexible when a consumer of messages. When used in a company that uses non-.Net technologies, NSB should be allowed to be easily bent to conform what producers dictate via conventions, not the other way around.

![quoterequested](https://f.cloud.github.com/assets/2625918/1614859/d9a01b3a-55f0-11e3-89e3-6939b2abe3b1.png)

**Comment by johnsimons** [comment]
@andreasohlund should this be moved to the requirements repo ?

**Comment by andreasohlund** [comment]
@gsogol According to this:

 https://github.com/Particular/NServiceBus.RabbitMQ/blob/message-id-strategy/src/NServiceBus.RabbitMQ.AcceptanceTests/When_using_a_custom_message_id_strategy.cs#L39

content type isn't needed. Can  you confirm?

**Comment by andreasohlund** [comment]
@gsogol it also seems like correlation id isn't needed? (can you confirm this as well)

**Comment by andreasohlund** [comment]
> should this be moved to the requirements repo ?

I think most of this issue is fixed. Except for the message type beeing needed. I talked to @SimonCropp and we belive we can adjust at least the xml serializer to use the root node in the xml to determine message type and if that is true we should be able to close this one

**Comment by gsogol** [comment]
Confirming all. I'm assuming you're talking about which handler to dispatch to based on the message type. Bottom line, is you want someone to use xsd.exe, generate classes in.net and java based on an xsd. nothing else should be required of the teams in either side of publishing and subscribing. Thank you for doing this. It's huge.

**Comment by andreasohlund** [comment]
With the upcoming 2.1.0 release the following message can be consumed without any headers (just a plain manual publish from the Rabbit management ui):

```
<?xml version="1.0"?>
<MyRequest xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns="http://tempuri.net/RawMsmqIntegration">
</MyRequest>
```

Message defined as:

```
 class MyRequest : IMessage
{
}
```

For this to work I used the following config:

```
   config.UseTransport<RabbitMQTransport>()
                .CustomMessageIdStrategy(m=>Guid.NewGuid().ToString());

```

Closing this issue
